runtime/near-vm-runner/src/wasmer2_runner.rs

```diff
@@ -40,76 +40,103 @@ const WASMER_FEATURES: wasmer::Features = wasmer::Features {
     fn fits_memory(&self, offset: u64, len: u64) -> bool {
-        match offset.checked_add(len) {
-            None => false,
-            Some(end) => self.0.size().bytes() >= Bytes(end as usize),
-        }
+        self.data_offset(offset)
+            .and_then(|(_, remaining)| {
+                let len = usize::try_from(len).ok()?;
+                Some(len <= remaining)
+            })
+            .unwrap_or(false)
     }
 
     fn read_memory(&self, offset: u64, buffer: &mut [u8]) {
-        let offset = offset as usize;
-        for (i, cell) in self.0.view()[offset..(offset + buffer.len())].iter().enumerate() {
-            buffer[i] = cell.get();
+        unsafe {
+            let memory = self.get_memory_buffer(offset, buffer.len());
+            // SAFETY: we verified indices into are valid and the pointer will always be valid as
+            // well. Our runtime is currently only executing Wasm code on a single thread, so data
+            // races aren't a concern here.
+            std::ptr::copy_nonoverlapping(memory, buffer.as_mut_ptr(), buffer.len());
         }
     }
 
     fn read_memory_u8(&self, offset: u64) -> u8 {
-        self.0.view()[offset as usize].get()
+        unsafe { *self.get_memory_buffer(offset, 1) }
     }
 
     fn write_memory(&mut self, offset: u64, buffer: &[u8]) {
-        let offset = offset as usize;
-        self.0.view()[offset..(offset + buffer.len())]
-            .iter()
-            .zip(buffer.iter())
-            .for_each(|(cell, v)| cell.set(*v));
+        unsafe {
+            let memory = self.get_memory_buffer(offset, buffer.len());
+            // SAFETY: we verified indices into are valid and the pointer will always be valid as
+            // well. Our runtime is currently only executing Wasm code on a single thread, so data
+            // races aren't a concern here.
+            std::ptr::copy_nonoverlapping(buffer.as_ptr(), memory, buffer.len());
+        }
     }
 }
 
-impl IntoVMError for wasmer::InstantiationError {
+impl IntoVMError for InstantiationError {
     fn into_vm_error(self) -> VMError {
         match self {
-            wasmer::InstantiationError::Link(e) => {
+            InstantiationError::Link(e) => {
                 VMError::FunctionCallError(FunctionCallError::LinkError { msg: e.to_string() })
             }
-            wasmer::InstantiationError::CpuFeature(e) => {
+            InstantiationError::CpuFeature(e) => {
                 panic!("host does not support the CPU features required to run contracts: {}", e)
             }
-            wasmer::InstantiationError::Start(e) => e.into_vm_error(),
-            wasmer::InstantiationError::HostEnvInitialization(_) => {
-                VMError::FunctionCallError(FunctionCallError::CompilationError(
-                    CompilationError::PrepareError(PrepareError::Instantiate),
-                ))
-            }
+            InstantiationError::Start(e) => e.into_vm_error(),
         }
     }
 }
```

Comment:
Mostly unrelated to the PR, but consider inlining this impl into a single place where it is called.